subgrounds/subgrounds.py

```diff
@@ -174,187 +172,147 @@ def mk_request(self, fpaths: FieldPath | list[FieldPath]) -> DataRequest:
+        return cast(DataResponse, transformer.send(data_resp))
 
     def execute_iter(
         self,
         req: DataRequest,
-        pagination_strategy: Optional[Type[PaginationStrategy]] = LegacyStrategy,
-    ) -> Iterator[dict[str, Any]]:
+        pagination_strategy: Type[PaginationStrategy] | None = LegacyStrategy,
+    ) -> Iterator[DocumentResponse]:
         """Same as `execute`, except that an iterator is returned which will iterate
         the data pages.
 
         Args:
-          req (DataRequest): The :class:`DataRequest` object to be executed
-          pagination_strategy (Optional[Type[PaginationStrategy]], optional): A Class
-            implementing the :class:`PaginationStrategy` ``Protocol``. If ``None``, then
-            automatic pagination is disabled. Defaults to :class:`LegacyStrategy`.
+          req: The :class:`DataRequest` object to be executed
+          pagination_strategy: A Class implementing the :class:`PaginationStrategy`
+           ``Protocol``. If ``None``, then automatic pagination is disabled.
+           Defaults to :class:`LegacyStrategy`.
 
         Returns:
-          Iterator[dict]: An iterator over the reponse data pages
+          An iterator over the :class:`DocumentResponse` pages.
+
+        ⚠️ DOES NOT apply global transforms across multiple documents or their pages.
+         Since we yield each page as we get it, it's not possible to accurately perform
+         the transforms since we don't collect the pages. This means transforms
+         expecting multiple documents or pages of documents will be inaccurate.
         """
 
-        def execute_document(doc: Document) -> Iterator[dict[str, Any]]:
-            subgraph: Subgraph = next(
-                self.subgraphs.values() | where(lambda sg: sg._url == doc.url)
-            )
-            if pagination_strategy is not None and subgraph._is_subgraph:
-                yield from paginate_iter(
-                    subgraph._schema,
-                    doc,
-                    pagination_strategy=pagination_strategy,
-                    headers=self.headers,
-                )
-            else:
-                yield client.query(
-                    doc.url,
-                    doc.graphql,
-                    variables=doc.variables,
-                    headers=self.headers,
+        document_transforms = {
+            url: subgraph._transforms for url, subgraph in self.subgraphs.items()
+        }
+        transformer = apply_transforms(self.global_transforms, document_transforms, req)
+        strategy = normalize_strategy(pagination_strategy)
+
+        data_req = cast(DataRequest, next(transformer))
+        for doc in data_req.documents:
+            paginator = paginate(self.subgraphs[doc.url]._schema, doc, strategy)
+            paginated_doc = next(paginator)
+            while True:
+                resp = client.query(paginated_doc, headers=self.headers)
+
+                next(transformer)  # toss empty None
```

Comment:
This is being called for each document, is this expected? Don't think we have tests for multiple Documents in one DataRequest so not sure if this behavior is tested

Reply:
This is expected and I've added a test for atleast multiple paginated documents. Multiple documents in a single `DataRequest` I suppose isn't tested.